Chore: Add integration test suite

seanlanglands opened the pull request:
### Description

Introduces a framework for writing integration tests, which also supports unit testing.

For integration testing, a base test case class is provided that contains `set_up` and `tear_down` methods that future class tests can use to ensure each test is run with a new instantiation of the main `S3_Media_Sync` plugin class.

seanlanglands committed:
Add integration test suite

GaryJones commented on bin/install-wp-tests.sh:
```diff
@@ -0,0 +1,181 @@
+
+install_wp
+install_test_suite
+install_db
```
Looks like a missing clear linespace is needed at the end of this file (see the red marker).

GaryJones commented on composer.json:
```diff
@@ -1,15 +1,34 @@
+	},
+	"require-dev": {
+		"phpunit/phpunit": "^9.0",
+		"yoast/phpunit-polyfills": "^1.0",
```
```suggestion
```

Not needed, as wp-test-utils already [includes phpunit-polyfills](https://github.com/Yoast/PHPUnit-Polyfills/blob/134921bfca9b02d8f374c48381451da1d98402f9/composer.json#L30) as its own dependency.

GaryJones commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
```
```suggestion
	xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.6/phpunit.xsd"
```

GaryJones commented on composer.json:
```diff
@@ -1,15 +1,34 @@
+		}
+	},
+	"require-dev": {
+		"phpunit/phpunit": "^9.0",
```
```suggestion
		"phpunit/phpunit": "^9.6",
```

Since a lot of features and fixes were added in minor versions of 9.x, let's require the latest minor of v9 as the minimum.

GaryJones commented on tests/integration/class-wp-integration-test-case.php:
```diff
@@ -0,0 +1,63 @@
+			require_once dirname( __FILE__, 3 ) . '/vendor/autoload.php';
+		}
+
+		$this->s3_media_sync = S3_Media_Sync::init();
```
I don't think I'm a fan of this line.

Why? Two reasons.

Firstly, the `init()` call is attempting to create the class as a singleton pattern (even though the constructor is public), and while that is often sean as negative for production code, it seems even worse for test code; any changes made in one test may carry through to further tests, rather than keeping everything in isolation.

Secondly, as this is assigned to the public property and presumably called from each test class extending this test case, it's effectively double-globalised, which again could affect isolation.

Let's see what the constructor of the class being tested does:

https://github.com/Automattic/s3-media-sync/blob/97616c31dd92d7e26eda5393d4b11746b2737e1b/inc/class-s3-media-sync.php#L20-L22

So I look at this and think: "What if there's a test where I need to update the saved options _before_ I instantiate the class (directly, or through the `init()` call)? I wouldn't be able to, because it's all nailed down once the test has started due to the structure of the `set_up()` (or, at least, one would need to be careful _when_ the `parent::set_up()` is called).

I can see how an integration test suite may want to limit the number of `get_options()` calls to the database, but maybe that then becomes a code smell that the constructor of the S3_Media_Sync class shouldn't be making calls to the database just to set up a class - a separate PR should look at instantiating the class and dependency-injecting the retrieved options (a Settings data transfer object, for instance, which can be started and injected via a Factory class), which then allows flexibility of running unit tests against this class since the database retrieval can be mocked instead.

I'll leave this up to you to consider for this particular PR, but it's probably worth thinking and capturing notes into an issue even if you can't work on this at the moment.

GaryJones commented on tests/integration/class-wp-integration-test-case.php:
```diff
@@ -0,0 +1,63 @@
+	public function tear_down(): void {
+		parent::tear_down();
+
+		$this::set_private_property(
+			$this->s3_media_sync::class,
+			$this->s3_media_sync,
+			'instance',
+			null
+		);
```
If the above changes were made in terms of the simplified setup, this tear_down reset would seem un-needed.

seanlanglands committed:
Clean up and set minimum phpunit version to 9.6

seanlanglands commented:
Setting this PR to draft as it needs refactoring independent of any `S3_Media_Sync` class simplification.

seanlanglands commented:
Specific `S3_Media_Sync` test can be found here:
https://github.com/Automattic/s3-media-sync/tree/chore/class-s3-media-sync-tests

GaryJones commented:
Last force-push just rebased branch into `develop`. No other changes.

GaryJones committed:
chore: Update phpunit config

Configure PHPUnit for improved test coverage and reporting.

GaryJones committed:
test: Add integration tests for S3 Media Sync core functionality

Add comprehensive integration test suite covering:
- Hooks registration
- Settings validation
- Stream wrapper registration

The new test files include tests for:
- Conditional hook registration based on settings
- Settings validation and error handling
- Stream wrapper registration and initial setup

GaryJones committed:
chore: Use PSR-4 naming for tests

This makes the current referencing easier, and it's what is required for later versions of PHPUnit anyway.

GaryJones committed:
test: Enhance integration tests with Mockery for S3 client validation

- Add Mockery as a dev dependency for improved mocking
- Update SettingsTest to include S3 client validation
- Implement comprehensive StreamWrapperTest with mocked S3 client
- Add tear_down method to close Mockery mocks after tests

GaryJones commented:
As of f150110, the tests are passing, with 8 tests and 18 assertions.

GaryJones committed:
test: Add integration test for media upload to S3

- Create MediaUploadTest to validate media synchronization with S3
- Implement test cases for image and document uploads
- Use Mockery to mock S3 client and validate upload process
- Cover upload path generation and S3 object creation

GaryJones committed:
test: Improve SettingsTest with robust error and state management

- Add set_up method to clear existing settings and errors before tests
- Enhance test_invalid_settings_cause_admin_error_notice with error state reset
- Update tear_down method to clean up settings and errors after tests
- Ensure clean test environment by resetting global error state

GaryJones committed:
test: Add integration test for image editor functionality with S3 sync

- Create ImageEditorTest to validate image editing operations sync with S3
- Implement test cases for crop, resize, and rotate image operations
- Use Mockery to mock S3 client and WP_Image_Editor for comprehensive testing
- Verify image editor changes are correctly synchronized to S3

GaryJones committed:
test: Allow code coverage to be generated

GaryJones committed:
test: Add integration test for bulk media upload to S3

- Create BulkOperationsTest to validate bulk media synchronization
- Implement test cases for multiple file types and bulk uploads
- Use Mockery to mock S3 client and validate bulk upload process
- Cover file upload scenarios with different media types

GaryJones committed:
feat: Enhance S3 media upload error handling

- Improve error handling in `add_attachment_to_s3()` method
- Add robust exception handling for S3 upload failures
- Wrap different types of errors into S3 exceptions for consistent error reporting
- Add error code extraction for stream wrapper-related errors

GaryJones committed:
refactor: Improve test helper methods and mock generation

- Enhance TestCase with reusable methods for test file and upload creation
- Add create_mock_s3_client method to simplify S3 client mocking
- Implement create_mock_command for consistent command mocking
- Replace hardcoded test file and upload creation with new helper methods
- Standardize error simulation and mock generation across test cases

GaryJones committed:
test: Enhance MediaUploadTest with content verification and error handling

- Add content verification for S3 file uploads
- Implement test for upload error handling with S3 exceptions
- Create test method to validate file content and upload failures
- Improve test coverage for media upload scenarios

GaryJones committed:
test: Enhance ImageEditorTest with content verification and error handling

- Add content verification for edited images synced to S3
- Implement test method to validate file content after image editing
- Create test case for S3 upload error handling during image editor operations
- Improve test coverage for image editing and synchronization scenarios

GaryJones committed:
test: Enhance BulkOperationsTest with content verification and error handling

- Add content verification for files uploaded to S3 during bulk operations
- Implement test method to validate file content and S3 upload success
- Create test case for S3 upload error handling with mock client
- Improve test coverage for bulk media upload scenarios
- Verify error handling and PHP error triggering for failed uploads

GaryJones committed:
test: Add integration test for file deletion from S3

- Create FileDeleteTest to validate file deletion functionality
- Implement test cases for deleting files with different S3 bucket configurations
- Add test method to verify successful file deletion from S3
- Implement error handling test for S3 deletion failures
- Cover file deletion scenarios with various file paths and bucket settings

GaryJones committed:
ci: Add integration test run workflow

GaryJones committed:
chore: Rename tests subdirectory

GaryJones commented:
There are more tests to be added, before we can then start refactoring the production classes (and then simplifying the tests down), but this PR is at a suitable place to be reviewed. 

There's only one production file that's changed, and that's by adding a try...catch around an existing function call so that there is much better error handling (that can be tested for).

The WP-CLI command is currently excluded from the tested classes, as that extends a class from vip-go-mu-plugins that needs addresing.

Overall Code Coverage Report:
  2025-02-22 21:25:58

 Summary:
  Classes:  0.00% (0/2)
  Methods: 37.88% (25/66)
  Lines:   42.91% (251/585)

S3_Media_Sync
  Methods:  50.00% (10/20)   Lines:  48.12% ( 77/160)
S3_Media_Sync_Stream_Wrapper
  Methods:  32.61% (15/46)   Lines:  41.33% (174/421)

GaryJones committed:
test: Update PHPUnit configuration

- Remove random test execution order to stop testdox showing incorrectly
- Change test group configuration from include to exclude
- Modify group filtering to skip specific tests

GaryJones committed:
refactor: Improve settings rendering with view files

- Extract settings rendering to separate view files
- Define `S3_MEDIA_SYNC_FILE` constant for consistent file path references
- Add integration tests for UI rendering

GaryJones committed:
feat: Improve form accessibility with label and id attributes

- Add `label_for` parameter to settings registration in `class-s3-media-sync.php`
- Update view templates to include matching `id` attributes for form inputs
- Add integration test for object ACL select element rendering
- Enhance form input accessibility and usability

GaryJones committed:
chore: tidy up use statements in test files

GaryJones committed:
refactor: Inline settings rendering and remove view templates

- Replace separate view template files with direct rendering in methods
- Remove redundant view files for S3 settings inputs
- Update integration tests to work with new rendering approach
- Simplify settings rendering in `class-s3-media-sync.php`

This undoes most of an earlier change, as it's no longer needed now that the tests call the render methods directly.

GaryJones commented:
Code Coverage Report:
  2025-02-23 15:24:06

 Summary:
  Classes:  0.00% (0/2)
  Methods: 45.45% (30/66)
  Lines:   46.51% (280/602)

S3_Media_Sync
  Methods:  75.00% (15/20)   Lines:  59.89% (106/177)
S3_Media_Sync_Stream_Wrapper
  Methods:  32.61% (15/46)   Lines:  41.33% (174/421)
